### DISCUSSION TITLE
Race condition in Decorator class?

### DISCUSSION THREAD

**Discussion by nomakewan** [scripthookvdotnet/scripthookvdotnet#1365] [Q&A]
Hello, it's me again, and again I have a random crash happening with a specific user. In fact, it's from the same user who had previously reported the race condition crash that was fixed in #1349 ! But this time it's not crashing during the "waiting for the game to finish loading" but rather when we try to apply our custom Decorator to vehicles. While for time machines this has always worked just fine, we now have an optional element that applies a certain set of custom Decorators to NPC vehicles, and this appears to be causing a crash on his end (but not mine).

The crash is logged in Event Viewer as follows:
```
Application: GTA5.exe
Framework Version: v4.0.30319
Description: The process was terminated due to an unhandled exception.
Exception Info: System.AccessViolationException
   at SHVDN.NativeFunc.NativeCall()
   at SHVDN.NativeFunc.InvokeInternal(UInt64, UInt64*, Int32)
   at SHVDN.NativeFunc+NativeTaskPtrArgs.Run()
   at SHVDN.ScriptDomain.ExecuteTaskWithGameThreadTlsContext(SHVDN.IScriptTask)
   at SHVDN.NativeFunc.Invoke(UInt64, UInt64*, Int32)
   at GTA.Native.Function.Call[[System.Boolean, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]](GTA.Native.Hash, GTA.Native.InputArgument, GTA.Native.InputArgument, GTA.Native.InputArgument)
   at GTA.DecoratorInterface.SetBool(GTA.Entity, System.String, Boolean)
   at FusionLibrary.Decorator.SetBool(System.String, Boolean)
   at BackToTheFutureV.HoverVehicle.set_IsHoverBoosting(Boolean)
   at BackToTheFutureV.HoverVehicle..ctor(GTA.Vehicle)
   at BackToTheFutureV.HoverVehicle.GetFromVehicle(GTA.Vehicle)
   at BackToTheFutureV.HoverMode.HoverMode_Tick(System.Object, System.EventArgs)
   at SHVDN.Script.DoTick()
   at SHVDN.Script.MainLoop()
   at System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
   at System.Threading.ThreadHelper.ThreadStart()
```

The Decorator code we use to interface with SHVDN (to extend SHVDN's abilities and allow assigning other data types such as Vector3 to a Decorator) can be found here: [Fusion Library Decorator (Nightly SHVDN Ver)](https://github.com/nomakewan/FusionLibrary/blob/nightlyshvdn/Other/Decorator.cs)

BTTFV's HoverMode/Hover Vehicle code is here:
[HoverVehicle](https://github.com/bttfv/BackToTheFutureV/blob/nightlyshvdn/BackToTheFutureV/Utility/HoverVehicle.cs)
[HoverMode](https://github.com/bttfv/BackToTheFutureV/blob/nightlyshvdn/BackToTheFutureV/Utility/HoverMode.cs)

Originally this code was iterating every vehicle in the world and applying the decorator if it didn't already have it. I recently made a change to only run on vehicles within 500.0 of the player, and this reduced the incidence of crashes for this specific user but did not resolve the underlying issue (which is why he recently experienced the same crash despite the updated code). I also have the ability to disable this code entirely, and doing so did eliminate the crash completely for that user (since we were no longer applying Decorators, this makes sense).

Help as always greatly appreciated! Thank you!

**Comment by kagikn** [comment]
Good to see you! I thought I wrote how to call native decorators wrong, but if natives to set decorators can crash in specific users' games not everyone's, I should study how memory ordering in x64 is more… So I'll be better at how to choose a way to synchronize data between threads. Maybe disassembling the internal function of `DECOR_SET_BOOL` will help me resolve this issue though.
 
EDIT:  In x64 (AMD64), according to Chapter 9.2 in *[System Programming Guide](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-sdm.html)* of *Intel® 64 and IA-32 Architectures Software Developer's Manual* (325384-081US, September 2023), reads can be reordered with older writes. Reads can't be reordered with locked instructions or memory fences, however. Reads can't reordered be with older reads and writes can't be reordered with older writes, either.

**Comment by kagikn** [comment]
Since #1387 seems to have got resolved a lot if not completely resolved with 200388b5143a4bc2e57f66d20d49c0e36a1c0b0d, I guess it's a good time for the user you mentioned to retry with a [nightly.46](https://github.com/scripthookvdotnet/scripthookvdotnet-nightly/releases/tag/v3.6.0-nightly.46) or a newer one?

**Reply by nomakewan** [reply]
Sadly not. While it did at least allow the game to load--a huge improvement over crashing on initial load-in--the same crash occurs a little while later as the player is navigating. Presumably this is just because the proverbial can has been kicked further down the road. Screenshot of the event viewer result below

![hoverdeco](https://github.com/scripthookvdotnet/scripthookvdotnet/assets/11485516/c76823e5-39ea-4ce0-88d6-a9ce97ea0b56)

What's worse is that even if I have him disable the hover subsystem, he now gets a random error thrown when trying to load GTAV. We're not sure what's causing it, but we know that it didn't occur prior to nightly.46. Sometimes it happens, sometimes it doesn't. Screenshot below:

![corefailure](https://github.com/scripthookvdotnet/scripthookvdotnet/assets/11485516/6c99befd-1993-4221-b169-ae0c7ff0a2ac)

I apologize for the delay in responding; was a busy week.

**Reply by kagikn** [reply]
I'm sure this is probably same issue as #1387 now. I believe this shit is so hard to completely resolve due to how SHV ticks.
